mlrun/datastore/base.py

```diff
@@ -157,6 +158,176 @@ def upload(self, key, src_path):
+        """Check if a partition directory (year=.., month=.., etc.) is in the time range."""
+        from dateutil.relativedelta import relativedelta
+
+        partition_start = datetime.datetime(
+            year=year,
+            month=month or 1,
+            day=day or 1,
+            hour=hour or 0,
+            tzinfo=start_time.tzinfo,
+        )
+        partition_end = (
+            partition_start
+            + relativedelta(
+                years=1 if month is None else 0,
+                months=1 if day is None and month is not None else 0,
+                days=1 if hour is None and day is not None else 0,
+                hours=1 if hour is not None else 0,
+            )
+            - datetime.timedelta(microseconds=1)
+        )
+
+        if end_time < partition_start or start_time > partition_end:
+            return False
+        return True
+
+    @staticmethod
+    def _list_partitioned_paths(
+        base_path: str,
+        start_time: datetime.datetime,
+        end_time: datetime.datetime,
+        partition_level: str,
+        filesystem,
+    ):
+        paths = []
+
+        def list_partition_paths_helper(
+            paths: list[str],
+            start_time,
+            end_time,
+            current_path: str,
+            partion_level: str,
+        ):
+            timing = current_path.rsplit("/")[-1].split("=")[0]
+            if filesystem.isfile(current_path):
+                paths.append(current_path.rsplit("/", 1)[0].rstrip("/"))
+                return
+            elif timing and timing == partion_level:
+                paths.append(current_path.rstrip("/"))
+                return
+            directories = filesystem.ls(current_path, detail=True)
+            if len(directories) == 0:
+                return
+            for directory in directories:
+                current_path = directory["name"]
+                parts = [p for p in current_path.split("/") if "=" in p]
+                kwargs = {}
+                for part in parts:
+                    part = part.split("=")
+                    key = part[0]
+                    value = int(part[1]) if part[1].isdigit() else part[1]
+                    kwargs[key] = value
+                if DataStore._is_directory_in_range(start_time, end_time, **kwargs):
+                    list_partition_paths_helper(
+                        paths, start_time, end_time, current_path, partion_level
+                    )
+
+        list_partition_paths_helper(
+            paths, start_time, end_time, base_path, partition_level
+        )
+        for i in range(len(paths)):
+            paths[i] = DataStore._reconstruct_path_from_base_path(base_path, paths[i])
+        return paths
+
+    @staticmethod
+    def _reconstruct_path_from_base_path(base_path: str, returned_path: str) -> str:
+        base_path = base_path.rstrip("/")
+        clean_return = returned_path.strip("/")
+
+        prefix, _, base_suffix = base_path.partition(":///")
```

Comment:
`base_suffix` is never used. Isn't this just

```suggestion
        prefix = base_path.split(":///", 1)[0]
```
?